### PR TITLE
[#114] Swagger API generator

### LINKED ISSUE #114: Generate documentation for the Web API
# Clarification and motivation

We should generate some document that details which endpoints our API supports, what they expect and what they return.

I think we can use [servant-openapi3](https://hackage.haskell.org/package/servant-openapi3) to automatically generate Swagger docs. Let's give it a try.

# Acceptance criteria

We have documentation detailing the Web API interface.

### PR DESCRIPTION
**Pull request by DK318**
## Problem 
We don't have documentation for our `coffer Web API`.

## Solution 
Added executable `coffer-swagger-api` which generates
Swagger docs.

## Related issue(s)

- Resolves #114

Fixed #

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Review by sancho20021 — CHANGES_REQUESTED** [review]
I tried importing build `swagger.json` using [swagger website](https://editor.swagger.io/), and some errors were shown. Is that normal?
image

**Comment by DK318** [comment]
> I tried importing build swagger.json using [swagger website](https://editor.swagger.io/), and some errors were shown. Is that normal?

They are produced because of bad `ToSchema` instances. They would be removed after #112 is merged.

File: lib/Backend/Vault/Kv/Internal.hs
Comment (dcastro) [inline lib/Backend/Vault/Kv/Internal.hs:238]:
```suggestion
  deriving stock (Eq, Show)
  deriving newtype (FromHttpApiData, ToParamSchema)
```

File: lib/Coffer/Path.hs
Comment (dcastro) [inline lib/Coffer/Path.hs:140]: Hm this instance doesn't look right.

It's generating this schema:

```json
            "EntryPath": {
                "type": "object",
                "required": [
                    "unEntryPath"
                ],
                "pattern": "(/${pathSegment})+",
                "properties": {
                    "unEntryPath": {
                        "items": {
                            "$ref": "#/components/schemas/PathSegment"
                        },
                        "type": "array",
                        "minItems": 1
                    }
                }
            },
```

And `EntryPath` is not a json object, it's a json string. And indeed, the examples shown in the swagger editor are wrong as well:

image

Comment (dcastro) [reply lib/Coffer/Path.hs:140]: Note: as a result of fixing this, I don't think the instance `ToSchema PathSegment` will be needed anymore.

**Review by dcastro — CHANGES_REQUESTED** [review]

File: app/swagger-api/Main.hs
Comment (dcastro) [reply app/swagger-api/Main.hs:21]: If this works, we should add a pipeline step to make sure the committed swagger.json is up to date.
I'll write an issue for this later, I'm just leaving this comment here as a reminder for myself.

File: lib/Entry.hs
Comment (DK318) [reply lib/Entry.hs:63]: ghc parser is going weird if use lens `pattern`. Idk how to fix it

Comment (dcastro) [reply lib/Entry.hs:63]: Ooh right, yeah I think that's it.
I think we can get around it by qualifying it, e.g. `Schema.pattern`

File: lib/Entry.hs
Comment (sancho20021) [reply lib/Entry.hs:166]: Never mind

File: lib/CLI/Types.hs
Comment (sancho20021) [reply lib/CLI/Types.hs]: Ok

File: lib/Backend/Vault/Kv/Internal.hs
Comment (sancho20021) [reply lib/Backend/Vault/Kv/Internal.hs]: Changed

File: lib/Coffer/Path.hs
Comment (sancho20021) [reply lib/Coffer/Path.hs]: Changed

File: lib/BackendName.hs
Comment (sancho20021) [reply lib/BackendName.hs]: Deleted

File: lib/Entry.hs
Comment (sancho20021) [reply lib/Entry.hs:63]: Fixed

File: app/swagger-api/Main.hs
Comment (sancho20021) [reply app/swagger-api/Main.hs]: Added

File: app/swagger-api/Main.hs
Comment (sancho20021) [reply app/swagger-api/Main.hs]: Done

File: lib/Entry.hs
Comment (sancho20021) [reply lib/Entry.hs:65]: Found some (I hope all) record-lens mixtures and removed there record syntax

File: app/swagger-api/Main.hs
Comment (sancho20021) [reply app/swagger-api/Main.hs]: Hmm, the auto-generated json doesn't contain the license.

I think it would be convenient to automatically add the license. Do I need to just edit the `app/swagger-api/Main.hs` so that the license is concatenated with the generated `json` before writing to the file?

**Comment by sancho20021** [comment]
Ok now tests don't pass. Maybe the problem was fixed after https://github.com/serokell/coffer/issues/113 had been closed, so rebase on main will help. 

Firstly I will address the rest of the comments, and then rebase and try to fix the CI tests fails

File: lib/Coffer/Path.hs
Comment (sancho20021) [reply lib/Coffer/Path.hs:163]: Fixed the `ToSchema EntryPath`. But `ToSchema PathSegment` is required by `ToSchema Directory`

File: app/swagger-api/Main.hs
Comment (sancho20021) [reply app/swagger-api/Main.hs:15]: I rewritten the API types from `(EntryPath, EntryPath)` to `PairObject EntryPath`

**Comment by sancho20021** [comment]
I've tried to rebase, but I didn't manage to. This error appears after rebasing the first commit:

```
app/swagger-api/Main.hs:17:17: error:
    • No instance for (ToParamSchema Coffer.Path.EntryPath)
        arising from a use of ‘toOpenApi’
    • In the first argument of ‘(&)’, namely
        ‘toOpenApi (Proxy :: Proxy API)’
      In the first argument of ‘(&)’, namely
        ‘toOpenApi (Proxy :: Proxy API) & info . title .~ "Coffer Web API"’
      In the first argument of ‘(&)’, namely
        ‘toOpenApi (Proxy :: Proxy API) & info . title .~ "Coffer Web API"
           & info . version .~ "1.0"’
   |
17 |   let openApi = toOpenApi (Proxy :: Proxy API)
```
I think after the commits are approved, I can squash everything except two last commits, perhaps it will make the rebase easier.

File: app/swagger-api/Main.hs
Comment (dcastro) [reply app/swagger-api/Main.hs]: Yup, I've added this file to `.reuse/dep5`

**Review by dcastro — APPROVED** [review]

**Comment by dcastro** [comment]
Ok, I've added:
* Examples for most schemas
* Descriptions to most query parameters
* CORS support (so we can use the "Try it out" and "Execute" buttons in the Swagger UI)
* A link to the swagger docs in the readme
* A CI step to make sure `docs/swagger.json` is up-to-date

**Comment by dcastro** [comment]
@sancho20021 can you approve please? ^^

**Comment by dcastro** [comment]
The docs for this branch can be seen here: https://petstore.swagger.io/?url=https://raw.githubusercontent.com/serokell/coffer/dk318/%2523114-swagger-api/docs/swagger.json#/

**Comment by sancho20021** [comment]
> @sancho20021 can you approve please? ^^

@dcastro Sure! Just approve or do a review and approve?

**Comment by dcastro** [comment]
>  Just approve or do a review and approve?

@sancho20021 If you have the time, a review would be welcome as well :smile: if not, that's fine too.

**Review by sancho20021 — APPROVED** [review]
@dcastro Ok. I'm on vacation :smile: so I'll let myself just approve)

**Pull request merged**
